src/corehost/cli/test_fx_ver/test_fx_ver.cpp

```diff
@@ -0,0 +1,195 @@
+    { _X("2.1.0"),                      { 2, 1, 0,  _X(""),                   _X("") }                 , false },
+    { _X("2.1.1"),                      { 2, 1, 1,  _X(""),                   _X("") }                 , false },
+    { _X("4.6.0-preview.19064.1"),      { 4, 6, 0,  _X("-preview.19064.1"),   _X("") }                 , false },
+    { _X("4.6.0-preview1-27018-01"),    { 4, 6, 0,  _X("-preview1-27018-01"), _X("") }                 , false },
+};
+
+const size_t cases = sizeof(orderedCases)/sizeof(TestCase);
+
+void checkPrecedence(size_t iVal, const fx_ver_t &iVer, size_t jVal, const fx_ver_t &jVer)
+{
+    if (iVal == jVal)
+    {
+        TEST_ASSERT( (iVer == jVer));
+        TEST_ASSERT(!(iVer <  jVer));
+        TEST_ASSERT(!(iVer >  jVer));
+        TEST_ASSERT( (iVer <= jVer));
+        TEST_ASSERT( (iVer >= jVer));
+        TEST_ASSERT(!(iVer != jVer));
+    }
+    else if (iVal < jVal)
+    {
+        TEST_ASSERT(!(iVer == jVer));
+        TEST_ASSERT( (iVer <  jVer));
+        TEST_ASSERT(!(iVer >  jVer));
+        TEST_ASSERT( (iVer <= jVer));
+        TEST_ASSERT(!(iVer >= jVer));
+        TEST_ASSERT( (iVer != jVer));
+    }
+    else
+    {
+        TEST_ASSERT(iVal > jVal);
+
+        TEST_ASSERT(!(iVer == jVer));
+        TEST_ASSERT(!(iVer <  jVer));
+        TEST_ASSERT( (iVer >  jVer));
+        TEST_ASSERT(!(iVer <= jVer));
+        TEST_ASSERT( (iVer >= jVer));
+        TEST_ASSERT( (iVer != jVer));
+    }
+}
+
+void checkPrecedence()
+{
+    size_t isame = 0;
+
+    for (size_t i = 0; i < cases; ++i)
+    {
+        fx_ver_t iver;
+        bool ivalid = fx_ver_t::parse(orderedCases[i].str, &iver);
+
+        TEST_ASSERT(ivalid);
+
+        if (orderedCases[i].same) isame++;
+
+        size_t jsame = 0;
+
+        for (size_t j = 0; j < cases; ++j)
+        {
+            fx_ver_t jver;
+            bool jvalid = fx_ver_t::parse(orderedCases[j].str, &jver);
+
+            TEST_ASSERT(jvalid);
+
+            if (orderedCases[j].same) jsame++;
+
+            checkPrecedence(i - isame, iver, j - jsame, jver);
+        }
+    }
+}
+
+void checkParsing(const TestCase &myCase)
+{
+    fx_ver_t ver;
+    bool valid = fx_ver_t::parse(myCase.str, &ver);
+
+    TEST_ASSERT(valid);
+    TEST_ASSERT(ver.get_major() == myCase.ver.major);
+    TEST_ASSERT(ver.get_minor() == myCase.ver.minor);
+    TEST_ASSERT(ver.get_patch() == myCase.ver.patch);
+    TEST_ASSERT(ver.is_prerelease() == !myCase.ver.pre.empty());
+    TEST_ASSERT(ver.as_str() == myCase.str);
+}
+
+void checkParsing()
+{
+    for (size_t i = 0; i < cases; ++i)
+    {
+        checkParsing(orderedCases[i]);
+    }
+}
+
+void checkInvalidVersions()
+{
+    pal::string_t invalidVersions[] =
+    {
+        _X(""),
+        _X("1"),
+        _X("1.1"),
+        _X("A.1.1"),
+        _X("1.A.1"),
+        _X("1.1.A"),
+        _X("1A.1.1"),
+        _X("1.1A.1"),
+        _X("1.1.1A"),
+        _X("1.1.1-"),
```

Comment:
Interesting - where in the product does this fail (just curious, since I would think the code would mark this as valid).

Reply:
It is caught by this code in `validIdentifiers`

```C++
    bool prerelease = ids[0] == _X('-');
    bool buildMeta = ids[0] == _X('+');

    if (!(prerelease || buildMeta))
    {
        // ids must start with '-' or '+' for prerelease & build respectively
        return false;
    }
```